accepted/2022/capability-analyzer/capability-analyzer.md

```diff
@@ -0,0 +1,205 @@
+    {
+        public CapabilityCheckAttribute(Type type, string propertyName);
+        public Type Type { get; }
+        public string PropertyName { get; }
+    }
+
+    [AttributeUsage(AttributeTargets.Method |
+                    AttributeTargets.Property |
+                    AttributeTargets.Field,
+                    AllowMultiple = true, Inherited = false)]
+    public sealed class CapabilityGuardAttribute : Attribute
+    {
+        public CapabilityGuardAttribute(Type assertedCapabilityAttributeType);
+        public Type AssertedCapabilityAttributeType { get; }
+    }
+}
+```
+
+The behavior is as follows:
+
+* Capabilities are represented as types extending from `CapabilityAttribute`
+* Capabilities should be prefixed with `Requires` followed by the capability,
+  such as `RequiresThreadCreation` or `RequiresDynamicCode`.
+* APIs that should only be used when this capability is available should be
+  marked with this attribute.
+* Capability attributes should indicate the type and member of the capability
+  check, which is expressed by applying the `[CapabilityCheck]` attribute on the
+  capability attribute itself. The referenced member can either be a method with
+  no parameters or a property. In either case the return type must be of type
+  `bool`. A value of `true` indicates that the capability exists. The
+  `[CapabilityCheck]` attribute can be applied multiple times, which would
+  indicate different ways the caller can check.
+* In addition, a consumer can introduce their own capability checks by marking
+  their APIs with the `[CapabilityGuard]` attribute. They reference the type of
+  capability attribute the guard is asserting. The guard can either be a method,
+  property, or a field. In case of methods, parameters are also permissible.
+  However, the return value must of be of type `bool`. A value of `true` implies
+  the capability is there. Custom guards are useful for cases where a feature
+  requires multiple checks. We recently added this to our [platform
+  compatibility analyzer][platform-compat] and it has proven to be invaluable
+  because it makes it much easier to enable code analysis for existing code
+  bases where capabilities are already checked as part of the infrastructure.
+
+### Threading Annotations
+
+In order to model the thread creation annotation, we need an attribute for the
+capability and a new capability API on `Thread`:
+
+```C#
+namespace System.Threading
+{
+    [AttributeUsage(AttributeTargets.All, AllowMultiple = false, Inherited = false)]
+    [CapabilityCheck(typeof(Thread), nameof(Thread.IsCreationSupported))]
+    public sealed class RequiresThreadCreationAttribute : CapabilityAttribute
+    {
+        public RequiresThreadCreationAttribute();
+    }
+
+    public partial class Thread
+    {
+        public static bool IsCreationSupported { get; }
+
+        [RequiresThreadCreation]
+        public void Start();
+    }
```

Comment:
This example makes sense to me, @terrajobst could you add example of `CapabilityGuardAttribute` usage?